### PR TITLE
feat:  LemonButton refactor

### PR DESCRIPTION
**Pull request by benjackwhite**
## Problem

Rework of buttons to match new Design System as closely as possible, drastically reducing the freedom people have with the LemonButton and therefore their ability to deviate from the design.

## Changes

Notable changes:
* Many options for LemonButton have been removed to simplify things
* "stealth" is now a status rather than a type as it is more todo with colors than anything else

Screenshot 2022-08-04 at 13 02 19
Screenshot 2022-08-04 at 13 02 32

* To get an idea of how messy things were before:
Before buttons where styled in a range of different ways, many combinations not really working at all...

👉 *Stay up-to-date with [PostHog coding conventions](https://posthog.com/docs/contribute/coding-conventions) for a smoother review.*

* Had to update Paths buttons and styles as they were super screwed up already
|Before|After|
|----|----|
|Screenshot 2022-08-04 at 11 03 35|Screenshot 2022-08-04 at 12 31 51|

* Slightly different button styles for playback of recordings

* Invite modal changes

* LemonSwitch changes (to fit design system)
@clarkus the padding feels weird to me but if you're happy then we can go with it (it exactly matches whats in Figma)

|Before|After|
|----|----|
|Screenshot 2022-08-05 at 12 16 34|Screenshot 2022-08-05 at 12 17 45|

## TODO

- [x] Fix buttons that are "stealth" but with a border (e.g. data range filter)
- [x] Fix buttons that are "stealth" but with status "danger" expecting to be red
- [x] Generally what to do about "stealth". It is essentially a color change (to default dark) and a font change. Maybe it should be a separate property?
- [x] Fix LemonSwitch
- [x] Fix Sidebar coffee icon
- [ ] Fix Sidebar hover icon background
- [x] Fix AuthorizedUrl edit view
- [ ]

### CODE REVIEW

**Comment by clarkus** [comment]
Some quick notes on what I'm seeing. Mostly thinking about the button sizing and padding. 

- One important note on padding - when an icon is present the horizontal padding between the icon and the button edge is reduced by 4px. Typically buttons are padded `padding: 8px 12px 8px 12px`. A icon placed to the left would reduce the left padding to 8px `padding: 8px 12px 8px 8px`. This is just to account of the extra spacing inserted by the viewbox. When icons are present on both sides, the padding would be `padding: 8px` on all sides. 
- The `line-height` of the text labels should be 24px, the same height of the icon viewbox. 
- What's the stealth variant for? I've never seen that in production or in figma. 
- I don't think we should specify disabled buttons. We don't really use them and they're a dark pattern we should avoid as much as possible. 
- Unsure a loading scenario is worth a separate entry - it's really a variation on the icon button just with an animated decoration. Your call on whether they should be distinct or not. 
- Buttons with icon decor placed to the right is almost always a triangle down to indicate a sub menu. 
- The button with the vertical delimiter is only used within insights right now - it's meant to separate a primary action from a secondary action to show a sub menu. Just calling it out as a limited and very specific use case. 
- We don't have buttons with right popouts - that is only in the navigation list. Maybe those are implemented as buttons, but I don't know if it makes sense to make that generally available as a pattern. It's just very specific to the project navigation. 
- Buttons don't usually have tooltips because they have clear labeling. Tooltips most often describe information icons or in some cases tags / labels.

**Comment by benjackwhite** [comment]
@clarkus 
- Padding will be fixed 👍 
- Line-height: I'll update it to 1.5rems (24px) but it will apply to the button, rather than the content as we have buttons elsewhere that have non-simple content
- Stealth: Stealth is used everywhere 😅 It is the same as the default blue button but with the text color as `text-default` and icons as `muted-alt` Just to name some example use cases: Sidebar buttons, Menu buttons, Buttons in Select element. There is around 130 occurrences in the code. 
- Disabled: We have disabled buttons in use on a lot of forms or elements where we have indicated that feature is locked - often with a Tooltip explaining why. Not sure what makes them a dark pattern? 
- LoadingIcon - also no strong opinion but it's definitely in use a lot so we have to leave it for now
- SideIcon - Can't say much to Figma but in the code the only decorative examples I could find (as in without sideAction or popover) are: `IconArrowDropDown (1) IconSettings(1) IconOpenInApp(1) IconUnfoldLess/More(2) IconOpenInNew(1)` 
- Tooltip: Not strong opinion here but if I remove it I see quite a few errors popup. We can mark it as deprecated but then we need to ensure all the cases where it is currently used have an alternative to work with
- Popups: The popup position isn't decided by the button, hence the Storybook is just demonstrating that the Button will pick the correct icon, depending on the stated position.

I think this is really the challenge of this design system. How to we translate all these rules and ideas about when and how certain things can be used into something clear for developers :D I guess everything in the comment above or in Figma needs to be translated clearly so that in usage of the component people know whether they're using it right without having to have all of these things told to them. Perhaps the Storybook should include all the descriptive text from Figma as well

**Comment by clarkus** [comment]
> Line-height: I'll update it to 1.5rems (24px) but it will apply to the button, rather than the content as we have buttons elsewhere that have non-simple content

Can you elaborate on this? What are some examples and are they currently captured in the figma examples? 

> Stealth: Stealth is used everywhere 😅 It is the same as the default blue button but with the text color as text-default and icons as muted-alt Just to name some example use cases: Sidebar buttons, Menu buttons, Buttons in Select element. There is around 130 occurrences in the code.

Confused on this - so it's used within other elements to contrast against their backgrounds? The examples in storybook show really low contrast text on a light background - just not understanding the use case for it. 

> Disabled: We have disabled buttons in use on a lot of forms or elements where we have indicated that feature is locked - often with a Tooltip explaining why. Not sure what makes them a dark pattern?

It's a dead end for a user with no immediate clear way to recover. It always forces a user to backtrack, figure out where they went wrong, and then to rework back to that thing they wanted to use. Just saying it's not a good pattern because it puts so much effort on the user when we could probably swap it for a component that has clear reasoning and steps to recover. While we have them now, we should try to avoid them as much as possible. 

> SideIcon - Can't say much to Figma but in the code the only decorative examples I could find (as in without sideAction or popover) are: IconArrowDropDown (1) IconSettings(1) IconOpenInApp(1) IconUnfoldLess/More(2) IconOpenInNew(1)

This might be a case where we look at specific instances? There is a good bit of inconsistency right now, so there could be scenarios where something is implemented in a way that diverges from the current design language. I just don't want to bake in those scenarios into the new work because they might be wrong. `IconArrowDropDown` is the most common case and has a very specific positioning. `IconSettings` should always be placed to the left within a button. If this is also applying to lists, I haven't been counting those as buttons but as special list item cases. `IconOpenInApp` not sure what this is. `IconUnfoldLess/More` these are placed to the left of the button but rarely are they in buttons... more often they are in lists. `IconOpenInNew` this is definitely a case to document.

**Comment by benjackwhite** [comment]
> > Line-height: I'll update it to 1.5rems (24px) but it will apply to the button, rather than the content as we have buttons elsewhere that have non-simple content
> 
> Can you elaborate on this? What are some examples and are they currently captured in the figma examples?

We have a lot of use cases where the entire row is clickable and often has other Button functionality like a sideAction (Manage Subscription list as an example) where the content is for example multiple lines of text. Another example is the homepage buttons. If we restrict the content to only being Text then this works but we currently use the Button for a much wider range of use cases.

|||
|----|----|
|Screenshot 2022-08-05 at 08 44 11|Screenshot 2022-08-05 at 08 44 48|

Why is it important to you that the text content has the line-height rather than the wrapping div?

> 
> > Stealth: Stealth is used everywhere 😅 It is the same as the default blue button but with the text color as text-default and icons as muted-alt Just to name some example use cases: Sidebar buttons, Menu buttons, Buttons in Select element. There is around 130 occurrences in the code.
> 
> Confused on this - so it's used within other elements to contrast against their backgrounds? The examples in storybook show really low contrast text on a light background - just not understanding the use case for it.
> 

I don't know the history of the decision but that's what is used in most Popover menus, Select popovers, Sidebar etc. 
Anywhere you see a button that when you hover over has a blue background but that has standard text (like the examples) i listed - that's "stealth" style.

> > Disabled: We have disabled buttons in use on a lot of forms or elements where we have indicated that feature is locked - often with a Tooltip explaining why. Not sure what makes them a dark pattern?
> 
> It's a dead end for a user with no immediate clear way to recover. It always forces a user to backtrack, figure out where they went wrong, and then to rework back to that thing they wanted to use. Just saying it's not a good pattern because it puts so much effort on the user when we could probably swap it for a component that has clear reasoning and steps to recover. While we have them now, we should try to avoid them as much as possible.
> 

Cool. So maybe a good way forward is whenever you find a disabled button in the design / implementation, you can call it out and propose the alternative and then over time we should be able to phase it out completely? I can try and mark it as deprecated but it would be good to indicate to the developer how they should work around it. (maybe that's a good thing to have in the Design System - "Don't do this - do this instead")

> > SideIcon - Can't say much to Figma but in the code the only decorative examples I could find (as in without sideAction or popover) are: IconArrowDropDown (1) IconSettings(1) IconOpenInApp(1) IconUnfoldLess/More(2) IconOpenInNew(1)
> 
> This might be a case where we look at specific instances? There is a good bit of inconsistency right now, so there could be scenarios where something is implemented in a way that diverges from the current design language. I just don't want to bake in those scenarios into the new work because they might be wrong. `IconArrowDropDown` is the most common case and has a very specific positioning. `IconSettings` should always be placed to the left within a button. If this is also applying to lists, I haven't been counting those as buttons but as special list item cases. `IconOpenInApp` not sure what this is. `IconUnfoldLess/More` these are placed to the left of the button but rarely are they in buttons... more often they are in lists. `IconOpenInNew` this is definitely a case to document.

I guess the question here is - do we want to allow Devs to specify the right-hand icon? If the answer is "no not ever" then we should get rid of it entirely. If the answer is "yes but only these icons" then we need to agree on what they are. And if it is "yes but not as a standard API for the button" then we need to accept that the styling, format etc. might not be easily enforceable when people try and add an icon specified in Figma but without the Button helping them do that?

**Comment by benjackwhite** [comment]
@clarkus regarding disabled state - one other way I see it is used is to disable a button whilst waiting for something (like loading of data. We might not want the user to be able to click a button whilst something related to the item is in progress. Again not commenting on whether that's correct, but just that it is another edge case to consider

**Comment by pauldambra** [comment]
Good guidance here on disabled state and accidental double click of buttons  https://design-system.service.gov.uk/components/button/

TL;DR

Disabled state -> have one but try and avoid it
Double click avoidance in the UI -> yes but also needs handling in the backend

File: frontend/src/scenes/ingestion/frameworks/CodeSnippet.scss [hunk 1]
```diff
@@ -4,13 +4,13 @@
...
         top: 0.75rem;
-        right: 0.75rem;
+        right: 0.25rem;
         gap: 0.5rem;
         font-size: 1.5rem;
-        .CodeSnippet__copy-button {
-            background: rgb(39, 40, 34) !important;
-            box-shadow: 0 0 10px rgb(39 40 34) !important;
-        }
+        // .CodeSnippet__copy-button {
```
Comment (pauldambra) [inline frontend/src/scenes/ingestion/frameworks/CodeSnippet.scss:10]: commented instead of deleted?

File: frontend/src/scenes/insights/filters/FunnelExclusionsFilter.tsx [hunk 2]
```diff
@@ -98,8 +98,7 @@ function ExclusionRowSuffix({
             <div style={{ flex: 1 }} />
             <LemonButton
                 icon={<IconDelete />}
-                type="alt"
-                // size="small"
+                status="primary-alt" // size="small"
```
Comment (pauldambra) [inline frontend/src/scenes/insights/filters/FunnelExclusionsFilter.tsx:101]: we don't have the size="small" comment on other usages. Should we remove it here?

File: frontend/src/styles/utilities.scss [hunk 3]
```diff
@@ -574,3 +574,10 @@
 .opacity-100 {
     opacity: 1;
 }
+
+.pointer-events-none {
+    pointer-events: none;
+}
+// .pointer-events-auto {
```
Comment (pauldambra) [inline frontend/src/styles/utilities.scss:581]: commented?

**Review by pauldambra — COMMENTED** [review]
Had a pass through just looking at the text

File: frontend/src/scenes/actions/ActionEdit.tsx [hunk 4]
```diff
@@ -232,34 +230,37 @@ export function ActionEdit({ action: loadedAction, id, onSave, temporaryToken }:
                     </Field>
                 </div>
                 <div>
-                    <div style={{ marginTop: '2rem' }}>
-                        <h2 className="subtitle">Match groups</h2>
+                    <div className="my-2">
```
[the remark this answers is not in the document]
Comment (benjackwhite) [reply frontend/src/scenes/actions/ActionEdit.tsx:233]: Nope! Good catch

File: frontend/src/scenes/insights/filters/FunnelExclusionsFilter.tsx (hunk 2 above)
Comment (benjackwhite) [reply frontend/src/scenes/insights/filters/FunnelExclusionsFilter.tsx:101]: Oh yeah thats some find and replace gone wrong

**Comment by pauldambra** [comment]
# currently in prod

Screenshot 2022-08-05 at 11 43 45

# and is fixed in this PR

Screenshot 2022-08-05 at 11 44 19

**Comment by pauldambra** [comment]
Screenshot 2022-08-05 at 11 54 01

I think this is different to prod. And maybe only because I'm staring at everything so much

The font size in the add breakdown button is 0.4px bigger than the font in the button below

#FontDetective

**Review by pauldambra — APPROVED** [review]
Have been through the UI and only saw two differences

1) a fix
2) a .4px font difference

I'd bias for getting this in since it's getting to an unreviewable size and then fixing forwards if people see anything

**Comment by kappa90** [comment]
This button in the sidebar doesn't have a hover state for the + icon
Screenshot 2022-08-05 at 16 35 57

File: frontend/src/lib/components/icons.tsx [hunk 5]
```diff
@@ -512,411 +485,397 @@ export function IconCalculate(props: React.SVGProps<SVGSVGElement>): JSX.Element
...
 }
+
 function InsightIcon({
     background = 'var(--muted-alt)',
     noBackground = false,
     children,
-    style,
+    ...props
 }: InsightIconProps): JSX.Element {
+    console.log(children, props)
```
Comment (kappa90) [inline frontend/src/lib/components/icons.tsx:873]: Nasty log is nasty

**Comment by kappa90** [comment]
This X button now has a bigger padding than production, thus not having enough space when hovered
Screenshot 2022-08-05 at 16 39 12

**Comment by kappa90** [comment]
> This X button now has a bigger padding than production, thus not having enough space when hovered Screenshot 2022-08-05 at 16 39 12

Same at the bottom of the modal, the divider doesn't have the right spacing vs. the buttons
Screenshot 2022-08-05 at 16 40 32

**Comment by kappa90** [comment]
This is on me but it's clear that the input doesn't have a white background now
Screenshot 2022-08-05 at 16 44 11

File: frontend/src/lib/components/icons.tsx (hunk 5 above)
Comment (benjackwhite) [reply frontend/src/lib/components/icons.tsx:873]: omg great catch

**Comment by benjackwhite** [comment]
@kappa90 thanks for that. Fixed a bunch of them or got them into a good enough state to merge. I think Modals will need a similar deep dive to sort them all out as we have different patterns everywhere 😅

**Comment by benjackwhite** [comment]
Spoke with @pauldambra and we agreed we need to just get this in. It's definitely better than before even if there is still more work to do but we run the risk of more and more merge conflicts over time. I will merge on Monday and take any remaining comments / issues into new issues to fix as we go 😊

**Review by kappa90 — APPROVED** [review]
:godmode:
